contracts/replier/src/lib.rs

```diff
@@ -0,0 +1,151 @@
+use cosmwasm_schema::{cw_serde, QueryResponses};
+use cosmwasm_std::{
+    entry_point, from_json, to_json_binary, to_json_vec, Binary, Deps, DepsMut, Env, MessageInfo,
+    QueryResponse, Reply, ReplyOn, Response, StdError, StdResult, SubMsg, WasmMsg,
+};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+#[cw_serde]
+pub struct InstantiateMsg {}
+
+#[cw_serde]
+pub struct ExecuteMsg {
+    msg_id: u8,
+    set_data_in_exec_and_reply: bool,
+    return_order_in_reply: bool,
+    exec_error: bool,
+    reply_error: bool,
+    reply_on_never: bool,
+    messages: Vec<ExecuteMsg>,
+}
+
+#[cw_serde]
+#[derive(QueryResponses)]
+pub enum QueryMsg {}
+
+pub const CONFIG_KEY: &[u8] = b"config";
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq, JsonSchema)]
+pub struct State {
+    pub order: Vec<u8>,
+}
+
+#[entry_point]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    _msg: InstantiateMsg,
+) -> StdResult<Response> {
+    deps.storage
+        .set(CONFIG_KEY, &to_json_vec(&State { order: vec![] })?);
+    Ok(Response::new())
+}
+
+#[entry_point]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    _info: MessageInfo,
+    msg: ExecuteMsg,
+) -> StdResult<Response> {
+    let data = deps.storage.get(CONFIG_KEY).unwrap();
+    let mut config: State = from_json(data)?;
+    if msg.msg_id <= 1 {
+        config.order.clear();
+    }
+    config.order.extend([0xEE, msg.msg_id]);
+    deps.storage.set(CONFIG_KEY, &to_json_vec(&config)?);
+
+    let mut resp = Response::new();
+
+    if msg.set_data_in_exec_and_reply {
+        resp = resp.set_data(Binary::new(vec![0xEE, msg.msg_id]));
+    }
+
+    if msg.exec_error {
+        return Err(StdError::generic_err(format!(
+            "Err in exec msg_id: {}",
+            msg.msg_id
+        )));
+    }
+
+    for next_msg in msg.messages {
+        let wasm_msg = WasmMsg::Execute {
+            contract_addr: env.contract.address.to_string(),
+            msg: to_json_binary(&next_msg).unwrap(),
+            funds: vec![],
+        };
+        let mut msg_id: u64 = msg.msg_id.into();
```

Comment:
Is it intentional that this uses the parent's `msg_id` field instead of `next_msg.msg_id`?

Reply:
That's a good question. Short answer is: yes, it was intentional. I double-checked the test scenarios in wasmd and it seems that it doesn't matter which of those two information is chosen for the reply ID. If we find a valid scenario where this will matter then we can change that in this contract.

Reply:
My main reason for asking is that it confused me a little while reading the [wasmd tests](https://github.com/CosmWasm/wasmd/pull/2052/files).
For example the first test result is: `0xee, 0x1, 0xee, 0x2, 0xee, 0x3, 0xbb, 0x2, 0xbb, 0x1, 0xee, 0x4, 0xee, 0x5, 0xbb, 0x4, 0xbb, 0x1`
There's two times `0xbb, 0x1` in there, once for message 2 and once for message 4, but it's not immediately clear which one was when. But then in this case it makes sense that it would have to be 2 and then 4, since 4 wasn't even executed when the first `0xbb, 0x1` came in, so I guess it's fine.

Reply:
Yes, the order of the calls is one thing. We should also consider scenarios where `data` field in resopnses are tested, for e.g. `0xa, 0x6, 0xa, 0x2, 0xee, 0x3, 0xbb, 0x2, 0xbb, 0x1`. Here we can see that we got a response path from the execute of Contract3, through reply of Contract2 and Contract1. It may be even possible that the msg ID in the reply entrypoint is a reduntant information, but I haven't had time to analyze this detail.

Reply:
I assumed that it's a message header. I guess there's some protobuf definition for the response data field?